### PR TITLE
Add support for RSA cryptograms

### PR DESCRIPTION
**Pull request by adsr**
cc #5 

Apple's spec is here: https://developer.apple.com/library/content/documentation/PassKit/Reference/PaymentTokenJSON/PaymentTokenJSON.html

### CODE REVIEW

**Comment by jimdoescode** [comment]
@adsr this looks good, my only comment is maybe that we add PHP constants to define ECC or RSA

**Comment by peteward** [comment]
Apologies for the slow reply, it's taking a while to find time to look at this.

At the latest I should have feedback next week.

File: applepay.c [hunk 1]
```diff
@@ -577,12 +603,31 @@ static int _applepay_verify_signature(applepay_state_t *state) {
...
+        if (NULL == (leaf_pub_key = X509_get_X509_PUBKEY(state->leaf_cert))) {
+            rc = APPLEPAY_ERROR_COULD_NOT_GET_LEAF_PUBKEY;
+            break;
+        }
+        if (1 != X509_PUBKEY_get0_param(&pkalg, NULL, NULL, NULL, leaf_pub_key)) {
+            rc = APPLEPAY_ERROR_COULD_NOT_GET_LEAF_PUBKEY_ALGORITHM;
+            break;
+        }
+        if (OBJ_obj2nid(pkalg) != (state->type == APPLEPAY_TYPE_ECC ? NID_X9_62_id_ecPublicKey : NID_rsaEncryption)) {
+            rc = APPLEPAY_ERROR_LEAF_CERT_WRONG_PUBKEY_ALGORITHM;
```
Comment (adsr) [inline applepay.c]: `break;`

**Comment by gonzalovilaseca** [comment]
I'm getting the following error `undefined symbol: EVP_PKEY_CTX_set_rsa_oaep_md`, does this branch need to be compiled with `OpenSSL 1.0.2`?

**Comment by adsr** [comment]
@gonzalovilaseca Yeah I think that is a fairly new macro. We might be able to support earlier versions if we add an ifdef for it? If you care to test that, the define is here https://github.com/openssl/openssl/blob/52df25c/include/openssl/rsa.h#L131

**Comment by gonzalovilaseca** [comment]
@adsr Seems to work fine, we've been able to decrypt the token, I need to do some further testing but so far it looks great.

**Comment by tileandrewarifin** [comment]
@adsr I've been trying to decrypt with RSA support but am still running into the error `APPLEPAY_ERROR_COULD_NOT_INIT_SECRET_CTX`. While my payment token version is EC_v1, my merchant identity certificate is RSA_v1. I am not sure what to do.

**Comment by adsr** [comment]
@tileandrewarifin I think @adamelso @peteward ran into a similar issue. Maybe they can chime in.

**Comment by adsr** [comment]
@tileandrewarifin Unfortunately I don't have any experience with the iOS / PassKit side of Apple Pay. This might possibly be related? https://developer.apple.com/documentation/passkit/pkaddpaymentpassrequestconfiguration/1615930-initwithencryptionscheme?language=objc It's the only place I was able to find that mentions a choice between EC and RSA.

**Comment by gonzalovilaseca** [comment]
The problem @adamelso and @peteward were having is because they(we) were using the merchant keys and cer, those are only used for the SSL connection, for decoding the token you have to use just the payment cer and payment ECC keys. There's no RSA involved.

Maybe this clarifies: https://github.com/etsy/applepay-php/pull/7

**Pull request merged**